File: TaskBackend/src/main/java/com/example/TaskBackend/TaskController.java
package com.example.TaskBackend;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.*;
import java.util.List;

@RestController
@CrossOrigin(origins = "http://localhost:5173")
public class TaskController {

    @Autowired
    private TaskRepository repository;

    @GetMapping("/tasks")
    public List<Task> getAllTasks() {
        return repository.findAll();
    }

    @PostMapping("/tasks")
    public Task createTask(@RequestBody Task task) {
        return repository.save(task);
    }

    @PutMapping("/tasks/{id}")
    public Task updateTask(@PathVariable Long id, @RequestBody Task updatedTask) {
        Task existingTask = repository.findById(id).orElse(null);
        if (existingTask != null) {
            // Note: Kotlin generates 'getDone' for the 'done' field automatically
            // But sometimes for boolean it uses 'isDone' or 'getDone'.
            // If 'getDone()' turns red, try 'isDone()'.
            existingTask.setDone(updatedTask.getDone());
            return repository.save(existingTask);
        }
        return null;
    }

    @DeleteMapping("/tasks/{id}")
    public void deleteTask(@PathVariable Long id) {
        repository.deleteById(id);
    }
}

File: TaskBackend/src/main/kotlin/com/example/TaskBackend/TaskController.kt
package com.example.TaskBackend
import org.springframework.transaction.annotation.Transactional
import org.springframework.web.bind.annotation.*
import org.springframework.http.ResponseEntity
import org.springframework.http.HttpStatus

@RestController
@RequestMapping("/tasks")
@CrossOrigin(origins = ["http://localhost:5173"])
class TaskController(private val repository: TaskRepository) {

    // 1. Get All Tasks (One-line function!)
    @GetMapping
    fun getAllTasks(): List<Task> = repository.findAll()

    // 2. Create a Task
    @PostMapping
    fun createTask(@RequestBody task: Task): Task = repository.save(task)

    // 3. Get Task by ID
    @GetMapping("/{id}")
    fun getTaskById(@PathVariable id: Long): ResponseEntity<Task> {
        return repository.findById(id)
            .map { task -> ResponseEntity.ok(task) }
            .orElse(ResponseEntity.notFound().build())
    }
    @PutMapping("/{id}")
    fun updateTask(@PathVariable id: Long, @RequestBody updatedTask: Task): Task {
        println("FORCE UPDATING ID: $id TO STATUS: ${updatedTask.completed}")

        // 1. Run the Direct SQL Update (The Hammer)
        repository.markTaskAsComplete(id, updatedTask.completed)

        // 2. Since we bypassed the standard save, we manually return the updated object
        // so React can update the screen immediately.
        return updatedTask
    }
    // 5. Delete Task
    @DeleteMapping("/{id}")
    fun deleteTask(@PathVariable id: Long): ResponseEntity<Void> {
        return if (repository.existsById(id)) {
            repository.deleteById(id)
            ResponseEntity.noContent().build()
        } else {
            ResponseEntity.notFound().build()
        }
    }
}